### PR TITLE
feat: add fail_project_under and fail_any_model_under options

### LINKED ISSUE #55: Configurable acceptable project score
Hi

It would be good if we could set either 

- a minimum project score 
- minimum required severity

and have `dbt-score` return a non-zero return code so that we can stop CI pipelines. At the minute, only the inability to find a manifest file returns a non-zero return code. Duplicate rules raise an exception

e.g. for minimum required project score 

```
[tool.dbt-score]
min_required_project_score=7.5
```

or passed on command line `--min_required_project_score 7.5`

for maximum allowed severity

```
[tool.dbt-score]
min_required_severity=MEDIUM
```
or passed on command line `--min_required_severity MEDIUM`

Should it raise an exception or simply log a message and exit with non-zero return code ?

Regards
Rob

### PR DESCRIPTION
**Pull request by WobblyRobbly**
This pull request is for Issue #55 

Decisions
- `fail_project_under` and `fail_any_model_under` can be specified in the `pyproject.toml` or via command line
- to avoid making this a breaking change a default value of `0.0` for `fail_project_under` and `fail_any_model_under` is set
- the exit point for failing is in `cli.py` and that required the `Evaluation` instance to be returned from `lint_dbt_project`

Please let me know your thoughts of potential changes

Closes #55

### CODE REVIEW

File: tests/test_cli.py
Comment (druzhinin-kirill) [inline tests/test_cli.py:2]:
```suggestion

import logging
```

just a small style thing, not covered with our linter setup yet 😞

File: src/dbt_score/evaluation.py
Comment (druzhinin-kirill) [inline src/dbt_score/evaluation.py:53]:
```suggestion
        self.min_model_score: float = 10.0
```

File: src/dbt_score/cli.py
Comment (druzhinin-kirill) [inline src/dbt_score/cli.py:134]: There might be some edge cases where rounding will lead to nonsensical output, like:

Given:
```
min_model_score = 8.65
fail_any_model_under = 8.69
```

```
Individual model score 8.7 is less than `fail_any_model_under` setting of 8.69
``` 

We can consider using only raw/rounded values (both in calculation and output) or making the rounding strategy a bit more advanced.

I do not think this should block this PR but it's definitely something to keep in mind 🙌

**Review by druzhinin-kirill — COMMENTED** [review]
Thanks for your contribution!
Left some tiny comments, already looks good 🚀

File: tests/test_cli.py
Comment (WobblyRobbly) [reply tests/test_cli.py:2]: This has been correccted

File: src/dbt_score/evaluation.py
Comment (WobblyRobbly) [reply src/dbt_score/evaluation.py:53]: This has been corrected

File: src/dbt_score/cli.py
Comment (WobblyRobbly) [reply src/dbt_score/cli.py:134]: That makes sense. I left the rounding the same as the rounding in the actual scoring.

File: src/dbt_score/evaluation.py
Comment (matthieucan) [inline src/dbt_score/evaluation.py:73]: I'm not sure we should calculate the minimum here, I think there'd be more value giving a list of all models which fall below the minimum allowed instead?

File: src/dbt_score/config.py
Comment (matthieucan) [inline src/dbt_score/config.py:73]: I would be in favour of a breaking change and have reasonable defaults here, e.g. `5.0`

File: src/dbt_score/evaluation.py
Comment (WobblyRobbly) [reply src/dbt_score/evaluation.py:73]: How would something like this be ?

```🥉 customers_view (score: 7.3)
    WARN (medium) dbt_score.rules.generic.columns_have_description: Columns lack a description: c_name, c_comment.
    OK   dbt_score.rules.generic.has_description
    WARN (medium) dbt_score.rules.generic.has_owner: Model lacks an owner.
    OK   dbt_score.rules.generic.public_model_has_example_sql
    OK   dbt_score.rules.generic.sql_has_reasonable_number_of_lines

🥈 customer_nationalities (score: 9.3)
    OK   dbt_score.rules.generic.columns_have_description
    OK   dbt_score.rules.generic.has_description
    OK   dbt_score.rules.generic.has_owner
    WARN (low) dbt_score.rules.generic.public_model_has_example_sql: The model description does not include an example SQL query.
    OK   dbt_score.rules.generic.sql_has_reasonable_number_of_lines

Project score: 8.3 🥈
2024-06-17 17:55:00,214 ERROR [dbt_score.cli] Model model.view_test.customers_view scored 7.3 which is less than `fail_any_model_under` setting of 10.0
2024-06-17 17:55:00,214 ERROR [dbt_score.cli] Model model.view_test.customer_nationalities scored 9.3 which is less than `fail_any_model_under` setting of 10.0
```

Comment (matthieucan) [reply src/dbt_score/evaluation.py:73]: Not sure I would use the logger, the output should match what's above to be efficiently parsed by humans.
I suggest to simplify it a bit:
```
Error: fail_any_model_under = 10.0
Model model.view_test.customers_view scored 7.3
Model model.view_test.customer_nationalities scored 9.3
```
and
```
Error: fail_project_under = 10.0
```

What do you think?

File: docs/configuration.md
Comment (jochemvandooren) [inline docs/configuration.md:53]:
```suggestion
  will fail with return code 1.
```

File: docs/configuration.md
Comment (jochemvandooren) [inline docs/configuration.md:55]:
```suggestion
  fail with return code 1.
```

File: src/dbt_score/cli.py
Comment (jochemvandooren) [inline src/dbt_score/cli.py:90]:
```suggestion
    help="Fail if any model is under this value.",
```

File: src/dbt_score/cli.py
Comment (jochemvandooren) [inline src/dbt_score/cli.py:83]:
```suggestion
    help="Fail if the project score is under this value.",
```

File: src/dbt_score/cli.py
Comment (jochemvandooren) [inline src/dbt_score/cli.py:147]: What about making it more in line with the failed models output?

```suggestion
            print(f"Error: fail_project_under = {config.fail_project_under}")
            print(f"Project scored {round(evaluation.project_score.value,1)}")
```

**Review by jochemvandooren — COMMENTED** [review]
Super nice feature :rocket:, works very well. I've added some nitpicky comments, just to align with the style we have :sweat_smile: 

Looks good to me! Just missing an entry in the CHANGELOG, could you add it? :bowing_man:

**Review by druzhinin-kirill — APPROVED** [review]
pre-approve from my side 🚀

File: docs/configuration.md
Comment (WobblyRobbly) [reply docs/configuration.md:53]: amended

File: docs/configuration.md
Comment (WobblyRobbly) [reply docs/configuration.md:55]: amended

File: src/dbt_score/cli.py
Comment (WobblyRobbly) [reply src/dbt_score/cli.py:90]: amended

File: src/dbt_score/cli.py
Comment (WobblyRobbly) [reply src/dbt_score/cli.py:83]: amended

**Comment by WobblyRobbly** [comment]
@jochemvandooren as we've added a Breaking Change i.e. will fail if the default model score isn't met (5.0) this should bump the version to 1.0.0 based on ?

```
## [1.0.0] - 2024-06-19

- Add `project_fail_under` configuration 
- Add `fail_any_model_under` configuration
- **breaking:** - default values of `5.0` for `project_fail_under` and `fail_any_model_under` will cause command to exit return code 1
```

**Comment by matthieucan** [comment]
> @jochemvandooren as we've added a Breaking Change i.e. will fail if the default model score isn't met (5.0) this should bump the version to 1.0.0 based on ?
> 
> ```
> ## [1.0.0] - 2024-06-19
> 
> - Add `project_fail_under` configuration 
> - Add `fail_any_model_under` configuration
> - **breaking:** - default values of `5.0` for `project_fail_under` and `fail_any_model_under` will cause command to exit return code 1
> ```

We're following semantic versioning, meaning if the major version is `0`, a breaking change will bump the minor version.
However no need for this to happen in the changelog file for this PR, this is done when releasing :)

File: src/dbt_score/cli.py
Comment (matthieucan) [inline src/dbt_score/cli.py:148]: I think it'd make sense to have this output done by `src/dbt_score/formatters/human_readable_formatter.py`. The `Formatter` class doesn't know about the config,  but it could easily receive it upon instantiation in `__init__`. I'm happy to take care of this if you prefer - let me know.

Comment (WobblyRobbly) [reply src/dbt_score/cli.py:148]: @matthieucan I've pushed the other PR fixes but happy for you to fix this. I'm struggling for time this week

Comment (matthieucan) [reply src/dbt_score/cli.py:148]: @WobblyRobbly Unfortunately it seems I can't push to this PR, despite the "Maintainers are allowed to edit this pull request" option. Is there some branch protection on your side?

Comment (matthieucan) [reply src/dbt_score/cli.py]: Thanks, that worked!

File: src/dbt_score/cli.py
Comment (jochemvandooren) [inline src/dbt_score/cli.py:142]: Not a big fan of using `x` personally, `score` would be more readable IMO

File: docs/configuration.md
Comment (jochemvandooren) [inline docs/configuration.md:52]: Btw, I think adding the default value would be useful here! (5.0)

**Review by jochemvandooren — APPROVED** [review]
Very nice, approved! :raised_hands: Two comments, which I will leave up to one of you!

File: src/dbt_score/cli.py
Comment (matthieucan) [reply src/dbt_score/cli.py:142]: That makes the line too long, and then spread over multiple lines I think we lose in readability

**Comment by matthieucan** [comment]
@WobblyRobbly Thank you so much for your contribution! First external PR that we merged for this project 🚀 🥇 
Looking forward to more of the same!

**Comment by WobblyRobbly** [comment]
Thanks for all your help on this one !!

**Pull request merged**
